code/model2.py

```diff
@@ -1,9 +1,129 @@
+    """
+
+    def __init__(self, epochs=10):
+        self.epochs = epochs
+        super(Model2, self).__init__()
+        # embedding lookup for 150 users each have 50 dimension representation
+        self.embedding_layer = nn.Embedding(150, 50)
+        # first hidden layer, linear layer with weights 200x500
+        # this should be the size of <sender+receiver representation>
+        self.h1_layer = nn.Linear(200, 500)
+        # ReLU activation used
+        self.relu = nn.ReLU()
+        # final linear layer that outputs the predicted email representation
+        self.email_layer = nn.Linear(500, 50)
+
+    def forward(self, s_id, r_ids,pv_emb, nv_emb):
+        """
+        Input is an integer sender_id and a list of receiver_ids. These ids are first converted into torch Variables and
+        then are used to lookup embeddings. If there are multiple receivers, an average embeddings is taken of all the
+        receivers. The gradients in the receiver embeddings are distributed accordingly.
+        Does a forward pass and returns the predicted email representation
+        :param s_id:
+        :param r_id
+        :return: email representation
+        """
+        # convert integers to long tensors
+        s_id = autograd.Variable(torch.LongTensor([s_id]))
+        r_ids = autograd.Variable(torch.LongTensor(r_ids))
+        # extract the embedding for all the receivers and take an average
+        r_emb = torch.mean(self.embedding_layer(r_ids), 0, True)
+        # extract the sender embedding
+        s_emb = self.embedding_layer(s_id)
+        # print s_emb.size()
+        # print r_emb.size()
+        #print size(pv_emb)
+        p_emb = torch.autograd.Variable(torch.from_numpy(pv_emb))
+        p_emb = p_emb.unsqueeze(0)
+        n_emb = torch.autograd.Variable(torch.from_numpy(nv_emb))
+        n_emb = n_emb.unsqueeze(0)
+        # print n_emb.size()
+        # simple concatenation of sender and receiver embedding
+        # sr_emb = torch.cat((s_emb, r_emb), 1)
+        # pn_emb = torch.cat((p_emb,n_emb), 1)
+        # print sr_emb.size()
+        # print pn_emb.size()
+        srpn_emb = torch.cat([s_emb, r_emb,p_emb,n_emb], 1)
+        # srpn_emb = torch.cat((srp_emb, n_emb),1)
+        # print srpn_emb.size()
+        h1 = self.relu(self.h1_layer(srpn_emb))
+        word_reps = self.email_layer(h1)
+        return word_reps
+
+    def get_average_rep(self, word_reps):
+        """
+        Assumes that word_reps is a numpy 2d array with every row as vector representation of word.
+        Calculates the mean across all rows to get average email representation
+        :param word_reps:
+        :return:
+        """
+        return np.mean(word_reps, axis=0)
+
     def train(self, emails, w2v):
-        pass
+        loss_criteria = nn.MSELoss()
+        optimizer = optim.RMSprop(self.parameters(), lr=0.001, alpha=0.6, momentum=0.6)
+
+        for epoch in range(self.epochs):
```

Comment:
A possible way to make this faster is to preprocess and keep the entire dataset of w_j-1 w_j+1 w_j etc ready
I don't know how much memory that would take but it would probably speed things up.

Reply:
I'm not sure how much of a gain this might have, since getting w_j-1, w_j+1 is just a lookup of embeddings from the word2vec model, right?